youtube_dlc/downloader/external.py

```diff
@@ -189,22 +202,85 @@ class Aria2cFD(ExternalFD):
-        cmd += ['--', info_dict['url']]
+        if 'url_list' in info_dict:
+            cmd += self._configuration_args(['--file-allocation', 'none', '--download-result', 'hide', '--uri-selector', 'inorder', '--console-log-level', 'warn', '-x16', '-j16', '-s16'])
+
+            url_list_file = '%s_urls.txt' % tmpfilename
+
+            url_list = []
+            for [i, url] in enumerate(info_dict['url_list']):
+                tmpsegmentname = '%s_%s.fragment' % (tmpfilename, i)
+                url_list.append('%s\n\tout=%s' % (url, tmpsegmentname))
+
+            with open(url_list_file, 'w') as f:
+                f.write('\n'.join(url_list))
+            cmd += ['-i', url_list_file]
+        else:
+            cmd += ['--', info_dict['url']]
         return cmd
 
+    def _call_downloader(self, tmpfilename, info_dict):
+        """ Either overwrite this or implement _make_cmd """
+        tn = os.path.basename(tmpfilename)
+
+        file_list = []
+        if 'url_list' in info_dict:
+            for [i, url] in enumerate(info_dict['url_list']):
+                tmpsegmentname = '%s_%s.fragment' % (tn, i)
+                file_list.append(tmpsegmentname)
+            info_dict['file_list'] = file_list
+
+        cmd = [encodeArgument(a) for a in self._make_cmd(tmpfilename, info_dict)]
+
+        self._debug_cmd(cmd)
+
+        p = subprocess.Popen(
+            cmd, stderr=subprocess.PIPE)
+        _, stderr = process_communicate_or_kill(p)
+        if p.returncode != 0:
+            self.to_stderr(stderr.decode('utf-8', 'replace'))
+
+        if 'url_list' in info_dict:
+            dn = os.path.dirname(tmpfilename)
+            with open(tmpfilename, 'wb') as dest:
+                for i in file_list:
+                    if 'decrypt_info' in info_dict:
+                        decrypt_info = info_dict['decrypt_info']
+                        with open(os.path.join(dn, i), 'rb') as src:
+                            if decrypt_info['METHOD'] == 'AES-128':
+                                iv = decrypt_info.get('IV')
+                                decrypt_info['KEY'] = decrypt_info.get('KEY') or self.ydl.urlopen(
+                                    self._prepare_url(info_dict, info_dict.get('_decryption_key_url') or decrypt_info['URI'])).read()
+                                encrypted_data = src.read()
+                                decrypted_data = AES.new(
+                                    decrypt_info['KEY'], AES.MODE_CBC, iv).decrypt(encrypted_data)
+                                dest.write(decrypted_data)
+                    else:
+                        shutil.copyfileobj(open(os.path.join(dn, i), 'rb'), dest)
+            for x in file_list:
+                try:
+                    file_path = os.path.join(dn, x)
+                    os.remove(file_path)
+                except OSError as e:
+                    print("Error: %s : %s" % (dn, e.strerror))
+            os.remove('%s_urls.txt' % tmpfilename)
```

Comment:
I think this should respect `--keep`. What do you think?

Reply:
Adjusted